src/foam/nanos/pm/FoldReducePMLogger.java

```diff
@@ -6,84 +6,118 @@
-    return new HashMap();
-  }
+    HashMap<String, PMInfo> initState = new HashMap<>();
+
+    // Load capture enabled PMInfos into initial state
+    if ( captureEnabledList_ != null && captureEnabledList_.size() > 0 ) {
+      for ( String key : captureEnabledList_.keySet() ) {
+        PMInfoId pmid = captureEnabledList_.get(key);
+
+        PMInfo pmi = new PMInfo();
+        pmi.setKey(pmid.getKey());
+        pmi.setName(pmid.getName());
+        pmi.setCapture(true);
 
+        initState.put(key, pmi);
+      }
+    }
+
+    return initState;
+  }
 
   /** Template method to Fold a new update into a state. **/
   public void fold(Object state, Object op) {
-    Map map = (Map) state;
+    Map<String, PMInfo> map = (Map<String, PMInfo>) state;
     PM  pm  = (PM) op;
 
     // Regular PMInfo
-    Object key = pm.getKey() + ":" + pm.getName();
-    PMInfo pi  = (PMInfo) map.get(key);
-
-    if ( pi == null ) {
-      pi = new PMInfo();
-      pi.setKey(pm.getKey());
-      pi.setName(pm.getName());
-      map.put(key, pi);
+    String key = pm.getKey() + ":" + pm.getName();
+    PMInfo pmi  = map.get(key);
+
+    if ( pmi == null ) {
+      pmi = new PMInfo();
+      pmi.setKey(pm.getKey());
+      pmi.setName(pm.getName());
+      map.put(key, pmi);
     }
 
-    pi.fold(pm);
-    if ( pi.getCapture() ) capture(pi);
+    pmi.fold(pm);
+    if ( pmi.getCapture() ) {
+      capture(pmi);
+    }
   }
 
+  public void capture(PMInfo pmi) {
+    StringBuilder trace = new StringBuilder("// Captured at: ");
+    trace.append(new Date());
+    trace.append(System.lineSeparator());
 
-  public void capture(PMInfo pi) {
-    StringBuilder trace = new StringBuilder();
     for ( StackTraceElement j : Thread.currentThread().getStackTrace() ) {
       trace.append(j.toString());
-      trace.append(System.getProperty("line.separator"));
+      trace.append(System.lineSeparator());
     }
 
-    pi.setCapture(false);
-    pi.setCaptureTrace(trace.toString());
+    pmi.setCapture(false);
+    pmi.setCaptureTrace(trace.toString());
+    put(pmi);
   }
 
-
   /** Template method to Merge two states. **/
   public Object reduce(Object state1, Object state2) {
     Map<String,PMInfo> m1 = (Map<String,PMInfo>) state1;
     Map<String,PMInfo> m2 = (Map<String,PMInfo>) state2;
 
-    for ( PMInfo pi2 : m2.values() ) {
-      String key = pi2.getKey() + ":" + pi2.getName();
-      PMInfo pi1 = m1.get(key);
+    for ( PMInfo pmi2 : m2.values() ) {
+      String key = pmi2.getKey() + ":" + pmi2.getName();
+      PMInfo pmi1 = m1.get(key);
 
-      if ( pi1 == null ) {
-        m1.put(key, pi2);
+      if ( pmi1 == null ) {
+        m1.put(key, pmi2);
       } else {
-        pi1.reduce(pi2);
+        pmi1.reduce(pmi2);
       }
     }
 
     return state1;
   }
 
+  /** Manage keys for maintaining capture flag between state and LocalState **/
+  public void put(PMInfo pmi) {
+    if ( pmi == null ) return;
+
+    String key = pmi.getKey() + ":" + pmi.getName();
```

Comment:
Needs to be synchronized, but is a bit tricky because you need to sync on the outer class, not 'this'.  You could do: synchronized ( FoldReducePMLoger.this ) { } around the whole body of this method, or move the contents of this method to the outer class, synchronize the method, and then call it from here.